scheduler/pkg/kafka/pipeline/client.go

```diff
@@ -155,78 +157,199 @@ func getSubsriberIp() (string, error) {
-			logger.Info("Expected at most a single model version", "numVersions", len(event.Versions), "name", event.GetPipelineName())
-			continue
+			return err
 		}
 
-		if len(event.Versions) == 1 {
-			pv, err := pipeline.CreatePipelineVersionWithStateFromProto(event.Versions[0])
-			if err != nil {
-				logger.Warningf("Failed to create pipeline state for pipeline %s with %s", event.PipelineName, protojson.Format(event))
-				continue
-			}
-
-			logger.Debugf("Processing pipeline %s version %d with state %s", pv.Name, pv.Version, pv.State.Status.String())
-			pc.pipelineStatusUpdater.Update(pv)
-
-			_, err = pc.pipelineInferer.LoadOrStorePipeline(pv.Name, false, false)
-			logger.Debugf("Stored pipeline %s", pv.Name)
-			if err != nil {
-				logger.WithError(err).Errorf("Failed to store pipeline %s", pv.Name)
-				continue
-			}
-		} else {
-			logger.Debugf("Received event with no versions for pipeline %s", event.PipelineName)
-			err := pc.pipelineInferer.DeletePipeline(event.PipelineName, false)
-			if err != nil {
-				logger.WithError(err).Errorf("Failed to delete pipeline %s", event.PipelineName)
-			}
-		}
+		processor.handleEvent(event)
 	}
-	logger.Infof("Closing connection to scheduler")
-	defer func() {
+}
+
+func (pc *PipelineSchedulerClient) cleanup(stream scheduler.Scheduler_SubscribePipelineStatusClient) {
+	pc.ready.Store(false)
+	pc.logger.Info("Closing connection to scheduler")
+	if stream != nil {
 		_ = stream.CloseSend()
-	}()
-	return nil
+	}
+}
+
+func (ep *EventProcessor) handleEvent(event *scheduler.PipelineStatusResponse) {
+	switch event.Operation {
+	case pb.PipelineStatusResponse_PipelineDelete:
+		ep.handleDeletePipeline(event)
+	case pb.PipelineStatusResponse_PipelineCreate:
+		ep.handleCreateOrUpdatePipeline(event)
+	}
+}
+
+type EventProcessor struct {
+	client         *PipelineSchedulerClient
+	grpcClient     scheduler.SchedulerClient
+	subscriberName string
+	logger         *logrus.Entry
+}
+
+func (ep *EventProcessor) handleDeletePipeline(event *scheduler.PipelineStatusResponse) {
+	pv, err := pipeline.CreatePipelineVersionWithStateFromProto(event.Versions[0])
+	if err != nil {
+		ep.reportFailure(
+			chainer.PipelineUpdateMessage_Delete,
+			nil,
+			fmt.Sprintf("No existing pipeline %s to delete", event.PipelineName),
+			event.Timestamp,
+			nil,
+		)
+		return
+	}
+
+	err = ep.client.pipelineInferer.DeletePipeline(event.PipelineName, false)
+	if err != nil {
+		ep.reportFailure(
+			chainer.PipelineUpdateMessage_Delete,
+			pv,
+			fmt.Sprintf("Failed to delete pipeline %s", event.PipelineName),
+			event.Timestamp,
+			err,
+		)
+		return
+	}
+
+	message := fmt.Sprintf("Pipeline %s deleted", event.PipelineName)
+	ep.reportSuccess(chainer.PipelineUpdateMessage_Delete, pv, message, event.Timestamp)
+}
+
+func (ep *EventProcessor) handleCreateOrUpdatePipeline(event *scheduler.PipelineStatusResponse) {
+	pv, err := pipeline.CreatePipelineVersionWithStateFromProto(event.Versions[0])
```

Comment:
Could we still log a warning in case the event contains multiple versions? We don't want to silently ignore unexpected input.